File: CSharpCodeBase/dialogs/dialoguesystem.lua
local Class = require 'utils.hump.class'

local DialogueSystem = Class({init=function(self)
  self.says = {}
  self.answers = {}
  self.dialogueUnity = luanet.GameFacade.dialogueSystem
  self.dialogueUnity:SetAnswerCallback(function (id)
      self:OnAnswerCallback(id)
  end)
end})

function DialogueSystem:LoadDialogue(dialogue)
  self:ResetBuffer()
  self.current = coroutine.create(require(dialogue))
  self.isRunning = false
end

function DialogueSystem:Show(dialogue)
  self:LoadDialogue(dialogue)
  self:Step()
  self.dialogueUnity:Show(self.says[1], self.answers[1], self.answers[2], self.answers[3], self.answers[4])
  luanet.TimeUtils.SetTimeScale(0)
  GameController.inputService:PushFrame("dialogue")
end

function DialogueSystem:ResetBuffer()
  self.says = {}
  self.answers = {}
end

function DialogueSystem:Say(phrase)
  self.says[#self.says + 1] = phrase
end

function DialogueSystem:Answer(phrase)
  self.answers[#self.answers + 1] = phrase
end

function DialogueSystem:OnAnswerCallback(id)
  self:Step(id)
  if not self:IsFinished() then
    self.dialogueUnity:Show(self.says[1], self.answers[1], self.answers[2], self.answers[3], self.answers[4])
  else
    self.dialogueUnity:Hide()
    GameController.inputService:PopFrame()
    luanet.TimeUtils.SetTimeScale(1)
    self.current = nil
  end
end

function DialogueSystem:Update()
  if self.current ~= nil then
    local lookVector = GameController.inputService:GetLookValue("dialogue")
    lookVector.x = 0
    if lookVector:Length() > 0.3 then
      if not self.switch then
        if lookVector.z > 0 then
          self.dialogueUnity:SelectDown()
        else  
          self.dialogueUnity:SelectUp()
        end
      end
      self.switch = true
    else
      self.switch = false
    end
    
    if GameController.inputService:LeftButtonWasPressed("dialogue") then
      self.dialogueUnity:JoysticApprove()
    end
  end
end

function DialogueSystem:Step(data)
  self:ResetBuffer()
  if self.current ~= nil then
    if not self.isRunning then
      coroutine.resume(self.current, self) 
      self.isRunning = true
    else
      coroutine.resume(self.current, data) 
    end
    return coroutine.status(self.current) ~= 'dead'
  else
    return false
  end
end

function DialogueSystem:IsFinished()
  return self.current == nil or coroutine.status(self.current) == 'dead'
end

return DialogueSystem

File: CSharpCodeBase/dialogs/dialoguesystem.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class dialoguesystem {
 public void init(self);
   this.says = {}
   this.answers = {}
   this.dialogueUnity = luanet.GameFacade.dialogueSystem;
   this.dialogueUnity:SetAnswerCallback(function (id){
       self:OnAnswerCallback(id);
   })
 }})
 public void LoadDialogue(dialogue){
   self:ResetBuffer();
   this.current = coroutine.create(require(dialogue));
   this.isRunning = false;
 }
 public void Show(dialogue){
   self:LoadDialogue(dialogue);
   self:Step();
   this.dialogueUnity:Show(this.says[1], this.answers[1], this.answers[2], this.answers[3], this.answers[4]);
   luanet.TimeUtils.SetTimeScale(0);
   GameController.inputService:PushFrame("dialogue");
 }
 public void ResetBuffer(){
   this.says = {}
   this.answers = {}
 }
 public void Say(phrase){
   this.says[#this.says + 1] = phrase;
 }
 public void Answer(phrase){
   this.answers[#this.answers + 1] = phrase;
 }
 public void OnAnswerCallback(id){
   self:Step(id);
   if(not self:IsFinished()  ){ 
     this.dialogueUnity:Show(this.says[1], this.answers[1], this.answers[2], this.answers[3], this.answers[4]);
   }else{
     this.dialogueUnity:Hide();
     GameController.inputService:PopFrame();
     luanet.TimeUtils.SetTimeScale(1);
     this.current = null;
   }
 }
 public void Update(){
   if(this.current != null  ){ 
     var lookVector = GameController.inputService:GetLookValue("dialogue");
     lookVector.x = 0;
     if(lookVector:Length() > 0.3  ){ 
       if(not this.switch  ){ 
         if(lookVector.z > 0  ){ 
           this.dialogueUnity:SelectDown();
         }else{  
           this.dialogueUnity:SelectUp();
         }
       }
       this.switch = true;
     }else{
       this.switch = false;
     }
     
     if(GameController.inputService:LeftButtonWasPressed("dialogue")  ){ 
       this.dialogueUnity:JoysticApprove();
     }
   }
 }
 public void Step(data){
   self:ResetBuffer();
   if(this.current != null  ){ 
     if(not this.isRunning  ){ 
       coroutine.resume(this.current, self) ;
       this.isRunning = true;
     }else{
       coroutine.resume(this.current, data) ;
     }
     return coroutine.status(this.current) != "dead";
   }else{
     return false;
   }
 }
 public void IsFinished(){
   return this.current == null  ||  coroutine.status(this.current) == "dead";
 }
 }}